scheduler/reconcile.go

```diff
@@ -778,14 +696,158 @@ func (a *allocReconciler) computePlacements(group *structs.TaskGroup,
 			place = append(place, allocPlaceResult{
 				name:               name,
 				taskGroup:          group,
-				downgradeNonCanary: canaryState,
+				downgradeNonCanary: isCanarying,
 			})
 		}
 	}
 
 	return place
 }
 
+// computeReplacements either applies the placements calculated by computePlacements,
+// or computes more placements based on whether the deployment is ready for placement
+// and if the placement is already rescheduling or part of a failed deployment.
+// The input deploymentPlaceReady is calculated as the deployment is not paused, failed, or canarying
+// The following rules are applied:
+// * If the deployment is place ready, apply all placements and return
+// * If not place ready, check if any allocs have been lost
+// * If allocs have been lost, determine the number of replacements that are needed.
+// * Add placements to the result for the lost allocs.
+// * If there are failed allocations, inspect them.
+// * If the placement is rescheduling, and not part of a failed deployment, add
+//   to the place set, and add the previous alloc to the stop set.
+// * Return the number of allocs still needed.
+func (a *allocReconciler) computeReplacements(deploymentPlaceReady bool, desiredChanges *structs.DesiredUpdates,
+	place []allocPlaceResult, rescheduleNow, lost allocSet, underProvisionedBy int) int {
+
+	if deploymentPlaceReady {
+		desiredChanges.Place += uint64(len(place))
+		// This relies on the computePlacements having built this set, which in
+		// turn relies on len(lostLater) == 0.
+		a.result.place = append(a.result.place, place...)
+		a.markStop(rescheduleNow, "", allocRescheduled)
+		desiredChanges.Stop += uint64(len(rescheduleNow))
+
+		min := helper.IntMin(len(place), underProvisionedBy)
+		underProvisionedBy -= min
+	} else if !deploymentPlaceReady {
```

Comment:
This conditional is always true at this point, it can just be `} else {`

Reply:
+1